Refactored implementation of `iterparentnodeids` in `nodes`.

chuckyQ opened the pull request:
<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [ ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->
Refactored `iterparentnodeids` in the `nodes` module to simplify the implementation and improve readability.

RonnyPfannschmidt commented:
failures seem to relate tothe official deprecation of pkg-resources

do we have a performance compare of the old/new impl? i recall that this function had gotten some optimizations as its used for hook call matching

chuckyQ commented:
The original function is faster (0.65sec compared to 0.80sec).
That is pretty big hit for the new function (23% longer), but this is also on one million tests (100,000 * `len(tests)`) and the absolute time difference is 0.15sec. Below is the timing test I used. The tests came from `test_nodes.py::test_iterparentnodeids` in the `testing` folder. I commented out the `assert` so it would not affect timing, but it can be uncommented to verify both functions work.

```python
# Python 3.10.6
# Timing test for previous implementation and new implementation
# Average previous implementation time: 0.75sec
# Average new implementation time: 0.90sec

from typing import Iterator
from typing import Optional
from time import time

SEP = '/'

tests = (
        ("", [""]),
        ("a", ["", "a"]),
        ("aa/b", ["", "aa", "aa/b"]),
        ("a/b/c", ["", "a", "a/b", "a/b/c"]),
        ("a/bbb/c::D", ["", "a", "a/bbb", "a/bbb/c", "a/bbb/c::D"]),
        ("a/b/c::D::eee", ["", "a", "a/b", "a/b/c", "a/b/c::D", "a/b/c::D::eee"]),
        ("::xx", ["", "::xx"]),
        # / only considered until first ::
        ("a/b/c::D/d::e", ["", "a", "a/b", "a/b/c", "a/b/c::D/d", "a/b/c::D/d::e"]),
        # : alone is not a separator.
        ("a/b::D:e:f::g", ["", "a", "a/b", "a/b::D:e:f", "a/b::D:e:f::g"]),
        # / not considered if a part of a test name
        ("a/b::c/d::e[/test]", ["", "a", "a/b", "a/b::c/d", "a/b::c/d::e[/test]"]),
)


def iterparentnodeids_prev_impl(nodeid: str) -> Iterator[str]:
    """Return the parent node IDs of a given node ID, inclusive.
    For the node ID
        "testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_source"
    the result would be
        ""
        "testing"
        "testing/code"
        "testing/code/test_excinfo.py"
        "testing/code/test_excinfo.py::TestFormattedExcinfo"
        "testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_source"
    Note that / components are only considered until the first ::.
    """
    pos = 0
    first_colons: Optional[int] = nodeid.find("::")
    if first_colons == -1:
        first_colons = None
    # The root Session node - always present.
    yield ""
    # Eagerly consume SEP parts until first colons.
    while True:
        at = nodeid.find(SEP, pos, first_colons)
        if at == -1:
            break
        if at > 0:
            yield nodeid[:at]
        pos = at + len(SEP)
    # Eagerly consume :: parts.
    while True:
        at = nodeid.find("::", pos)
        if at == -1:
            break
        if at > 0:
            yield nodeid[:at]
        pos = at + len("::")
    # The node ID itself.
    if nodeid:
        yield nodeid


def iterparentnodeids_new_impl(nodeid: str) -> Iterator[str]:
    """Return the parent node IDs of a given node ID, inclusive.

    For the node ID

        "testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_source"

    the result would be

        ""
        "testing"
        "testing/code"
        "testing/code/test_excinfo.py"
        "testing/code/test_excinfo.py::TestFormattedExcinfo"
        "testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_source"

    Note that / components are only considered until the first ::.
    """
    # The root Session node - always present.
    yield ""

    if nodeid == '':
        return

    if nodeid.startswith('::'):
        # Weird case of '::x' (no path)
        yield nodeid
        return

    path, sep, node = nodeid.partition('::')

    sections = path.split(SEP)
    for i, _ in enumerate(sections, start=1):
        yield SEP.join(sections[:i])

    if sep == '':
        # Only a path
        return

    nodes = node.split('::')
    for i, _ in enumerate(nodes, start=1):
        yield '::'.join([path] + nodes[:i])


def stress_test(func, prompt):
    start = time()
    for _ in range(100_000):
        for node, expected in tests:
            got = list(func(node))
            # assert got == expected
    end = time()
    print(prompt, f'{end-start:0.2f} sec')


stress_test(iterparentnodeids_prev_impl, 'Previous Implementation:')
stress_test(iterparentnodeids_new_impl, 'New Implementation:')


```

bluetech commented:
Hi @chuckyQ, I somewhat prefer the existing version (which is to be expected as I'm the one who wrote the most recent version it...), and since it's also faster I prefer to keep it. But thanks for the suggestion!
